src/hash.cr

```diff
@@ -57,9 +57,30 @@ class Hash(K, V)
     value
   end
 
-  # See also: `Hash#fetch`.
+  # Returns the value for the key given by *key*.
+  # If not found, returns the default value given by `Hash.new`, otherwise raises `KeyError`.
+  #
+  # ```
+  # h = {"foo" => "bar"}
+  # h["foo"] # => "bar"
+  #
+  # h = Hash(String, String).new("bar")
+  # h["foo"] # => "bar"
+  #
+  # h = Hash(String, String).new { "bar" }
+  # h["foo"] # => "bar"
+  #
+  # h = Hash(String, String).new
+  # h["foo"] # raises KeyError
+  # ```
   def [](key)
-    fetch(key)
+    fetch(key) do
+      if (block = @block) && key.is_a?(K)
```

Comment:
Part below seems invalid to me:

```cr
if ... key.is_a?(K)
else
  raise KeyError.new "Missing hash key: #{key.inspect}"
end
```

Why wrong `key` type should trigger such error?

Reply:
Because it was so (look below). This code just moved.

And, I guess, because Ruby works like that:

> If the key can't be found, there are several options: With no other arguments, it will raise a KeyError exception;

https://devdocs.io/ruby~2.5/hash#method-i-fetch

Example: https://repl.it/@AlexWayfer/UnlinedGlossyQuotient

Reply:
> And, I guess, because Ruby works like that:

Ruby doesn't have strong typing, and so this specific edge-case.

Reply:
@Sija If the `is_a?(K)` check is not there, invoking the block won't compile. I added that check.

Reply:
@asterite Would that be _wrong thing to do_ ™? is this check is present inside of other `Hash` methods?

Reply:
I don't understand what you mean. There's an issue about restricting these methods and it's a PITA to use, so it as never done.

Reply:
I see, it looked odd to me to have two of those guards as a logic for handling the case of missing key, but your explanation makes total sense, thanks!